File: templates/problems/04_stack/04_generate_parentheses/go/generate_parentheses_test.go
package generate_parentheses

import (
	"sort"
	"testing"
)

func TestGenerateParenthesis(t *testing.T) {
	tests := []struct {
		name string
		n    int
		want []string
	}{
		{"n=1", 1, []string{"()"}},
		{"n=2", 2, []string{"(())", "()()"}},
		{"n=3", 3, []string{"((()))", "(()())", "(())()", "()(())", "()()()"}},
		{"n=4 count", 4, nil}, // 14 combinations
		{"n=0", 0, []string{""}},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			got := generateParenthesis(tt.n)
			if tt.want == nil {
				// just check count for larger n
				if tt.n == 4 && len(got) != 14 {
					t.Errorf("generateParenthesis(%d) returned %d results, want 14", tt.n, len(got))
				}
				return
			}
			sort.Strings(got)
			sort.Strings(tt.want)
			if len(got) != len(tt.want) {
				t.Errorf("generateParenthesis(%d) = %v, want %v", tt.n, got, tt.want)
				return
			}
			for i := range got {
				if got[i] != tt.want[i] {
					t.Errorf("generateParenthesis(%d) = %v, want %v", tt.n, got, tt.want)
					return
				}
			}
		})
	}
}

File: templates/problems/04_stack/04_generate_parentheses/typescript/generateParentheses.test.ts
import { describe, it, expect } from 'vitest';
import { generateParenthesis } from './generateParentheses';

describe('GenerateParentheses', () => {
  it('n=1', () => {
    expect(generateParenthesis(1).sort()).toEqual(['()']);
  });

  it('n=2', () => {
    expect(generateParenthesis(2).sort()).toEqual(['(())', '()()']);
  });

  it('n=3', () => {
    expect(generateParenthesis(3).sort()).toEqual([
      '((()))',
      '(()())',
      '(())()',
      '()(())',
      '()()()',
    ]);
  });

  it('n=4 count', () => {
    expect(generateParenthesis(4).length).toBe(14);
  });

  it('n=0', () => {
    expect(generateParenthesis(0)).toEqual(['']);
  });
});
